From db8bcd9b5cb0a72c1cc45d1ecefa32286a8cd7a7 Mon Sep 17 00:00:00 2001
From: anand <anekkunt@redhat.com>
Date: Thu, 29 Oct 2015 21:36:57 +0530
Subject: [PATCH 75/98] glusterd: fix info file checksum mismatch during upgrade

issue: probing a new  node(>=3.1.x)  from 3.0.x cluster is moving the peer to rejected state.

fix: Disperse vol support is added from 3.1 release, so write disperse fields (disperse_count=0
and redundancy_count=0) in vol info file  only if cluster version supported.

Downstream specific change:
      MACRO GD_OP_VERSION_3_6_0 is raplaced with GD_OP_VERSION_3_7_0 becasuse disperse vol support in added in downstream in 3.1 and op=version of
3.1 is GD_OP_VERSION_3_7_0.

Backport of:
>Change-Id: I11d5e2e337b9bbaddc8e52ca7295ba481beb1132
>BUG: 1276423
>Signed-off-by: anand <anekkunt@redhat.com>
>Reviewed-on: http://review.gluster.org/12464
>Reviewed-by: Pranith Kumar Karampuri <pkarampu@redhat.com>
>Tested-by: Gluster Build System <jenkins@build.gluster.com>
>Reviewed-by: Kaushal M <kaushal@redhat.com>
>Tested-by: NetBSD Build System <jenkins@build.gluster.org>

Change-Id: I2762f7fdda4b28bbe06dffce52854fd34334b660
BUG: 1276541
Signed-off-by: anand <anekkunt@redhat.com>
Reviewed-on: https://code.engineering.redhat.com/gerrit/60605
Reviewed-by: Atin Mukherjee <amukherj@redhat.com>
Tested-by: Atin Mukherjee <amukherj@redhat.com>
---
 xlators/mgmt/glusterd/src/glusterd-store.c |   25 +++++++++++++++----------
 1 files changed, 15 insertions(+), 10 deletions(-)

diff --git a/xlators/mgmt/glusterd/src/glusterd-store.c b/xlators/mgmt/glusterd/src/glusterd-store.c
index 57b3268..5e9c5f2 100644
--- a/xlators/mgmt/glusterd/src/glusterd-store.c
+++ b/xlators/mgmt/glusterd/src/glusterd-store.c
@@ -909,11 +909,14 @@ glusterd_volume_exclude_options_write (int fd, glusterd_volinfo_t *volinfo)
         char         buf[PATH_MAX]  = "";
         int32_t      ret            = -1;
         xlator_t    *this           = NULL;
+        glusterd_conf_t  *conf      = NULL;
 
         this = THIS;
         GF_ASSERT (this);
         GF_ASSERT (fd > 0);
         GF_ASSERT (volinfo);
+        conf = this->private;
+        GF_VALIDATE_OR_GOTO (this->name, (conf != NULL), out);
 
         snprintf (buf, sizeof (buf), "%d", volinfo->type);
         ret = gf_store_save_value (fd, GLUSTERD_STORE_KEY_VOL_TYPE, buf);
@@ -946,17 +949,19 @@ glusterd_volume_exclude_options_write (int fd, glusterd_volinfo_t *volinfo)
         if (ret)
                 goto out;
 
-        snprintf (buf, sizeof (buf), "%d", volinfo->disperse_count);
-        ret = gf_store_save_value (fd, GLUSTERD_STORE_KEY_VOL_DISPERSE_CNT,
-                                   buf);
-        if (ret)
-                goto out;
+        if (conf->op_version >= GD_OP_VERSION_3_7_0) {
+                snprintf (buf, sizeof (buf), "%d", volinfo->disperse_count);
+                ret = gf_store_save_value (fd, GLUSTERD_STORE_KEY_VOL_DISPERSE_CNT,
+                                           buf);
+                if (ret)
+                        goto out;
 
-        snprintf (buf, sizeof (buf), "%d", volinfo->redundancy_count);
-        ret = gf_store_save_value (fd, GLUSTERD_STORE_KEY_VOL_REDUNDANCY_CNT,
-                                   buf);
-        if (ret)
-                goto out;
+                snprintf (buf, sizeof (buf), "%d", volinfo->redundancy_count);
+                ret = gf_store_save_value (fd, GLUSTERD_STORE_KEY_VOL_REDUNDANCY_CNT,
+                                           buf);
+                if (ret)
+                        goto out;
+        }
 
         snprintf (buf, sizeof (buf), "%d", volinfo->version);
         ret = gf_store_save_value (fd, GLUSTERD_STORE_KEY_VOL_VERSION, buf);
-- 
1.7.1

